joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/DownloadDependenciesPass.scala

```diff
@@ -12,65 +12,85 @@ import org.slf4j.LoggerFactory
-            val futures = mod.dependencies
-              .filter(dep => dep.beingUsed)
-              .map(dependency => {
-                Future {
-                  val dependencyStr = s"${dependency.module}@${dependency.version}"
-                  val cmd           = s"go get $dependencyStr"
-                  val results       = synchronized(ExternalCommand.run(cmd, prjDir))
+      val projDir = tmpDir.pathAsString
+      parentGoMod
+        .getModMetaData()
+        .foreach(mod => {
+          ExternalCommand.run("go mod init joern.io/temp", projDir) match {
+            case Success(_) =>
+              mod.dependencies
+                .filter(dep => dep.beingUsed)
+                .map(dependency => {
+                  val cmd     = s"go get ${dependency.dependencyStr()}"
+                  val results = ExternalCommand.run(cmd, projDir)
                   results match {
                     case Success(_) =>
                       print(". ")
-                      processDependency(dependencyStr)
+                      processor.queue.put(Some(dependency))
                     case Failure(f) =>
                       logger.error(s"\t- command '$cmd' failed", f)
                   }
-                }
-              })
-            val allResults: Future[List[Unit]] = Future.sequence(futures)
-            Await.result(allResults, Duration.Inf)
-          case Failure(f) =>
-            logger.error("\t- command 'go mod init joern.io/temp' failed", f)
-        }
-      })
+                })
+            case Failure(f) =>
+              logger.error("\t- command 'go mod init joern.io/temp' failed", f)
+          }
+        })
+    }
+    processor.queue.put(None)
+    processorThread.join()
   }
 
-  private def processDependency(dependencyStr: String): Unit = {
-    val gopath             = Try(sys.env("GOPATH")).getOrElse(Seq(os.home, "go").mkString(JFile.separator))
-    val dependencyLocation = (Seq(gopath, "pkg", "mod") ++ dependencyStr.split("/")).mkString(JFile.separator)
-    File.usingTemporaryDirectory("godep") { astLocation =>
-      val depConfig = Config()
-        .withInputPath(dependencyLocation)
-        .withIgnoredFilesRegex(config.ignoredFilesRegex.toString())
-        .withIgnoredFiles(config.ignoredFiles.toList)
-      // TODO: Need to implement mechanism to filter and process only used namespaces(folders) of the dependency.
-      // In order to achieve this filtering, we need to add support for inclusive rule with goastgen utility first.
-      val astGenResult = new AstGenRunner(depConfig).execute(astLocation).asInstanceOf[GoAstGenRunnerResult]
-      val goMod = new GoModHelper(
-        Some(depConfig),
-        astGenResult.parsedModFile.flatMap(modFile => GoAstJsonParser.readModFile(Paths.get(modFile)).map(x => x))
-      )
-      new MethodAndTypeCacheBuilderPass(None, astGenResult.parsedFiles, depConfig, goMod, goGlobal).process()
+  private class DependencyProcessorQueue extends Runnable {
```

Comment:
@DavidBakerEffendi This dependency processor queue I created to start the processing of downloaded dependency in parallel to downloading other dependencies. Earlier, it was downloading one dependency process it and then pickup another dependency. 

NOTE: We cannot parallelise downloading dependency as we use `go get <depdency>` command, which can be executed one at a time.

Reply:
Right, we would have to download directly from the repository otherwise, but this is fine, we don't want to put too much strain on the dependency server either as some have DDoS protection